Generate OPENJSON with WITH unless ordering is required

roji opened the pull request:
~This PR based based on top of #30976 and #30956, review those first and ignore the first commits here.~

We currently generate SQL Server OPENJSON expressions without WITH, applying a regular relational cast to applying the element store type to the output:

```sql
WHERE [t].[Id] IN (
    SELECT CAST([i].[value] AS uniqueidentifier) AS [value]
    FROM OPENJSON(@__ids_0) AS [i]
```

We use this form because it allows preserving the ordering of the JSON array (not needed in the above specific query), and that's not possible with the OPENJSON form that accepts WITH.

This PR switches to using OPENJSON with WITH in contexts where preserving the ordering isn't needed, such as the above query:

```sql
WHERE [t].[Id] IN (
    SELECT [i].[value]
    FROM OPENJSON(@__ids_0) WITH ([value] uniqueidentifier '$') AS [i]
```

Where ordering is important, we continue to use OPENJSON without WITH as today.

Compared to OPENJSON with WITH, OPENJSON without WITH has some performance drawbacks (see https://github.com/dotnet/efcore/issues/13617#issuecomment-1556016939, https://github.com/dotnet/efcore/issues/13617#issuecomment-1561125154). In addition, some conversions from JSON are only possible when using the WITH variant (see https://github.com/dotnet/efcore/issues/30727#issuecomment-1561765456).

/cc @yv989c, @pfritschi

roji commented:
Converting this back to a draft - there are probably patterns where ordering should be preserved that aren't detected by the current implementation (e.g. projecting out the entire collection as-is, applying an aggregate function that cares about ordering, such as string.Join). It may be a better approach to start with OPENWITH without WITH (and with the ORDER BY), and later convert to the more efficient form (with WITH) only if the ORDER BY has been removed (i.e. as a natural result of the subquery being inside IN/EXISTS).

maumar commented on src/EFCore.SqlServer/Query/Internal/SqlServerQueryTranslationPostprocessor.cs:
```diff
@@ -81,4 +82,100 @@ private sealed class SkipWithoutOrderByInSplitQueryVerifyingExpressionVisitor :
+                    foreach (var column in openJsonExpression.ColumnInfos)
+                    {
+                        var typeMapping = _typeMappingSource.FindMapping(column.StoreType);
+                        Check.DebugAssert(typeMapping is not null, "typeMapping is not null");
```
fix assert message - if check fails it means typeMapping *is* null

roji replied:
Ah, I thought we were doing it this way everywhere (it's what the IDE completes for me automatically) - I read it like "you failed the assertion that X".

I don't really mind either way, the important thing is to be able to trace the exception to the right line etc.

maumar replied:
which IDE? 😠 ;) from what I saw we do what the problem is, i don't mind either way but we should be consistent.

roji replied:
You seem to be right... Will change.

maumar commented on src/EFCore.SqlServer/Query/Internal/SqlServerQueryableMethodTranslatingExpressionVisitor.cs:
```diff
@@ -490,19 +405,29 @@ protected override Expression VisitExtension(Expression expression)
         ///     doing so can result in application failures when updating to a new Entity Framework Core release.
         /// </summary>
         protected virtual TableValuedFunctionExpression ApplyTypeMappingsOnOpenJsonExpression(
```
also change return type

maumar commented on src/EFCore.SqlServer/Query/Internal/SqlServerQueryableMethodTranslatingExpressionVisitor.cs:
```diff
@@ -490,19 +405,29 @@ protected override Expression VisitExtension(Expression expression)
+            if (openJsonExpression.JsonExpression is not SqlParameterExpression { TypeMapping: null } parameterExpression)
             {
+                Check.DebugAssert(
+                    openJsonExpression.JsonExpression.TypeMapping is not null, "openJsonExpression.JsonExpression.TypeMapping is not null");
```
update message

roji commented:
@maumar new version up, rebased on latest main.

I added two tests per your offline comments:
* Column_collection_OrderByDescending_ElementAt applies an ordering to a (naturally ordered) column collection. This shows the natural ordering (by `key`) getting erased (e.g. x.OrderBy(a).OrderBy(b) removes the first ordering). But it did uncover a small bug in ElementAt, so thanks!
* Inline_collection_Join_ordered_column_collection joins against an ordered collection; here too, the ordering gets elided since the ordering of joined rowsets has no meaning.
